providers/okta/application.go

```diff
@@ -0,0 +1,252 @@
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&identifyResponse)
+	return
+}
+
+type IntrospectResponse struct {
+	Success *struct {
+		Name string
+		Href string
+	}
+}
+
+func (source ApplicationSAMLSource) Introspect(ctx context.Context, client *http.Client, prevURL *url.URL, stateToken StateToken) (introspectResponse IntrospectResponse, err error) {
+	// We need to run introspect to get our next link.
+	// Now use /idp/idx/identify to get the IDP link to follow.
+	introspectVals := map[string]any{"stateToken": stateToken}
+	introspectBytes, _ := json.Marshal(introspectVals)
+	introspectURL := prevURL.ResolveReference(&url.URL{Path: "/idp/idx/introspect"})
+	resp, err := client.Post(introspectURL.String(), "application/json", bytes.NewReader(introspectBytes))
+	if err != nil {
+		return
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&introspectResponse)
+	return
+}
+
+func extractStateToken(r io.Reader) (StateToken, error) {
+	// Extremely cursed
+	bodyBuf, err := io.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+
+	// Extremely cursed
+	matches := stateTokenExpr.FindSubmatch(bodyBuf)
+	idx := stateTokenExpr.SubexpIndex("Token")
+	if len(matches) < idx+1 {
+		return "", errors.New("no match found")
+	}
+
+	// The JavaScript in the response has HTML-encoded characters in.
+	// When a web browser reads it, this is no problem. But, when we read it, it is.
+	// Specifically the only character observed so far is \x2D which is a hyphen.
+	return StateToken(strings.ReplaceAll(string(matches[idx]), "\\x2D", "-")), nil
+}
+
+// GetAssertion attempts to retrieve a SAML assertion from the given source and returns the bytes of that assertion.
+//
+// These bytes are not parsed.
+func (source ApplicationSAMLSource) GetAssertion(ctx context.Context, username, password string) ([]byte, error) {
+	jar, _ := cookiejar.New(&cookiejar.Options{
+		PublicSuffixList: publicsuffix.List,
+	})
+	client := http.Client{Jar: jar}
+	req, _ := http.NewRequest("GET", string(source.URL()), nil)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("could not send initial application request: %w", err)
+	}
+
+	stateToken, err := extractStateToken(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("could not find state token: %w", err)
+	}
+
+	identifyResponse, err := source.Identify(ctx, &client, resp.Request.URL, username, password, stateToken)
+	if err != nil {
+		return nil, fmt.Errorf("could not call /idp/idx/identify: %w", err)
+	}
+
+	method, ok := DetermineUpgradePath(identifyResponse)
+	if !ok {
+		return nil, errors.New("could determine 2fa upgrade path - users probably does not have a supported device")
+	}
+
+	nextStateToken, err := method.Upgrade(ctx, &client)
+	if err != nil {
+		return nil, fmt.Errorf("could not upgrade session with mfa: %w", err)
+	}
+
+	// HACK: In the live version of Okta, a different series of events is followed, where the correct URL obtained by following /idp/idx/introspect.
```

Comment:
This hints at some lovecraftian level horror inside of DuoIframe's Upgrade method.